prov/mxm/src/mlxm.h

```diff
@@ -0,0 +1,306 @@
+#include <sys/socket.h>
+#include <netdb.h>
+#include <rdma/fabric.h>
+#include <rdma/fi_prov.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_tagged.h>
+#include <rdma/fi_rma.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_errno.h>
+#include <rdma/fi_log.h>
+#include <fi_list.h>
+#include <mxm/api/mxm_api.h>
+#include "mpool.h"
+#include "uthash.h"
+static int mlxm_errno_table[MXM_ERR_LAST] = {
+	0,			/* MXM_OK = 0, */
+	-FI_ENOMSG,		/* MXM_ERR_NO_MESSAGE, */
+	-EWOULDBLOCK,		/* MXM_ERR_WOULD_BLOCK, */
+	-FI_EIO,		/* MXM_ERR_IO_ERROR, */
+	-FI_ENOMEM,		/* MXM_ERR_NO_MEMORY, */
+	-FI_EINVAL,		/* MXM_ERR_INVALID_PARAM, */
+	-FI_ENETUNREACH,	/* MXM_ERR_UNREACHABLE, */
+	-FI_EINVAL,		/* MXM_ERR_INVALID_ADDR, */
+	-FI_ENOSYS,		/* MXM_ERR_NOT_IMPLEMENTED, */
+	-FI_EMSGSIZE,		/* MXM_ERR_MESSAGE_TRUNCATED, */
+	0,			/* MXM_ERR_NO_PROGRESS, */
+	-FI_ETOOSMALL,		/* MXM_ERR_BUFFER_TOO_SMALL, */
+	-FI_ENOENT,		/* MXM_ERR_NO_ELEM, */
+	-FI_EIO,		/* MXM_ERR_SOME_CONNECTS_FAILED, */
+	-FI_ENODEV,		/* MXM_ERR_NO_DEVICE, */
+	-FI_EBUSY,		/* MXM_ERR_BUSY, */
+	-FI_ECANCELED,		/* MXM_ERR_CANCELED, */
+	-FI_EINVAL,		/* MXM_ERR_SHMEM_SEGMENT, */
+	-EEXIST,		/* MXM_ERR_ALREADY_EXISTS, */
+	-FI_EINVAL,		/* MXM_ERR_OUT_OF_RANGE, */
+	-FI_ETIMEDOUT		/* MXM_ERR_TIMED_OUT, */
+        /* MXM_ERR_LAST */
+};
+
+static inline
+int mlxm_errno(int err)
+{
+        if (err>=0 && err < MXM_ERR_LAST)
+                return mlxm_errno_table[err];
+	else
+		return -FI_EOTHER;
+}
+
+
+
+#define MLXM_SUPPORTED_FLAGS ( FI_SEND | FI_RECV )
+#define MLXM_DEFAULT_FLAGS   (0)
+extern struct fi_provider mlxm_prov;
+
+typedef struct mlxm_cq_entry       mlxm_cq_entry_t;
+typedef struct mlxm_fid_cq         mlxm_fid_cq_t;
+typedef struct mlxm_req            mlxm_req_t;
+typedef struct mlxm_fid_domain     mlxm_fid_domain_t;
+typedef struct mlxm_cq_entry_queue mlxm_cq_entry_queue_t;
+typedef struct mlxm_fid_av         mlxm_fid_av_t;
+typedef struct mlxm_fid_ep         mlxm_fid_ep_t;
+typedef struct mlxm_fid_fabric     mlxm_fid_fabric_t;
+
+struct mlxm_cq_entry {
+        void				*ptr;
+	mxm_conn_h			src_addr;
+        mlxm_cq_entry_t 		*next;
+};
+
+struct mlxm_req {
+        union {
+		mxm_send_req_t	sreq;
+		mxm_recv_req_t	rreq;
+        } mxm_req;
+        struct mpool* pool;
+        uint16_t   mq_id;
+};
+
+
+struct mlxm_mq_entry {
+        uint16_t mq_key;
+        mxm_mq_h mq;
+        UT_hash_handle hh;
+};
+
+struct mlxm_mq_storage{
+        struct mlxm_mq_entry *hash;
+};
+
+struct mlxm_globals{
+        mxm_h mxm_context;
+        mxm_ep_h mxm_ep;
+        struct mlxm_mq_storage mq_storage;
+        struct mpool *req_pool;
+};
+extern struct mlxm_globals mlxm_globals;
+
+#define HASH_FIND_INT16(head,findint,out)                                        \
+    HASH_FIND(hh,head,findint,sizeof(uint16_t),out)
+#define HASH_ADD_INT16(head,intfield,add)                                        \
+    HASH_ADD(hh,head,intfield,sizeof(uint16_t),add)
```

Comment:
These macros look unnecessary.  In any case, I would replace uint16_t with int16_t to match the names.